components/layout/display_list_builder.rs

```diff
@@ -322,6 +329,47 @@ impl FragmentDisplayListBuilding for Fragment {
         }
     }
 
+    fn compute_background_image_size(&self,
+                                     style: &ComputedValues,
+                                     bounds: &Rect<Au>,
+                                     image: &png::Image)
+                                     -> Size2D<Au> {
+        // If `aspect_ratio` < 1.0, the image is tall; otherwise, it is wide.
+        let aspect_ratio = (image.width as f64) / (image.height as f64);
+        let intrinsic_size = Size2D(Au::from_px(image.width as int),
+                                    Au::from_px(image.height as int));
+        match (style.get_background().background_size.clone(), aspect_ratio < 1.0) {
```

Comment:
Still seems to be an issue: Rather than compare the image’s ratio to 1.0, you want to compare it to the ratio of the background positioning area.